rust/parquet/src/data_type.rs

```diff
@@ -583,32 +740,198 @@ pub(crate) mod private {
         fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
             self
         }
     }
 
+    // TODO - Why does macro importing fail?
+    /// Reads `$size` of bytes from `$src`, and reinterprets them as type `$ty`, in
+    /// little-endian order. `$ty` must implement the `Default` trait. Otherwise this won't
+    /// compile.
+    /// This is copied and modified from byteorder crate.
+    macro_rules! read_num_bytes {
+        ($ty:ty, $size:expr, $src:expr) => {{
+            assert!($size <= $src.len());
+            let mut buffer = <$ty as $crate::util::bit_util::FromBytes>::Buffer::default();
+            buffer.as_mut()[..$size].copy_from_slice(&$src[..$size]);
+            <$ty>::from_ne_bytes(buffer)
+        }};
+    }
+
     impl ParquetValueType for super::ByteArray {
-        fn encoded(&self) -> EncodedValue<'_> {
-            EncodedValue::Bytes { data: self.data() }
+        #[inline]
+        fn encode<W: std::io::Write>(values: &[Self], writer: &mut W, _: &mut BitWriter) -> Result<()> {
+            for value in values {
+                let len: u32 = value.len().try_into().unwrap();
+                writer.write_all(&len.to_ne_bytes())?;
+                let raw = value.data();
+                writer.write_all(raw)?;
+            }
+            Ok(())
+        }
+
+        #[inline]
+        fn set_data(decoder: &mut PlainDecoderDetails, data: ByteBufferPtr, num_values: usize) {
+            decoder.data.replace(data);
+            decoder.start = 0;
+            decoder.num_values = num_values;
+        }
+
+        #[inline]
+        fn decode(buffer: &mut [Self], decoder: &mut PlainDecoderDetails) -> Result<usize> {
+            let data = decoder.data.as_mut().expect("set_data should have been called");
+            let num_values = std::cmp::min(buffer.len(), decoder.num_values);
+            for i in 0..num_values {
+                let len: usize = read_num_bytes!(u32, 4, data.start_from(decoder.start).as_ref()) as usize;
+                decoder.start += std::mem::size_of::<u32>();
+
+                if data.len() < decoder.start + len {
+                    return Err(eof_err!("Not enough bytes to decode"));
+                }
+
+                let val: &mut Self = buffer[i]
+                    .as_mut_any()
+                    .downcast_mut()
+                    .unwrap();
+
+                val.set_data(data.range(decoder.start, len));
+                decoder.start += len;
+            }
+            decoder.num_values -= num_values;
+
+            Ok(num_values)
+        }
+
+        #[inline]
+        fn dict_encoding_size(&self) -> (usize, usize) {
+            (std::mem::size_of::<u32>(), self.len())
+        }
+
+        #[inline]
+        fn as_any(&self) -> &dyn std::any::Any {
+            self
         }
 
+        #[inline]
+        fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
+            self
+        }
+    }
+
+    impl ParquetValueType for super::FixedLenByteArray {
```

Comment:
The lifted `FixedLenByteArray` becomes used here to restore (actually make _better_) the performance of encoding and decoding.

The logic is I think different enough to warrant some duplication.